src/Megamix/Patches.cpp

```diff
@@ -94,29 +82,82 @@ namespace Megamix::Patches {
+        std::array<u16, 5> newRowIds { 0x101, 0x101, 0x101, 0x101, 0x101 };
+        size_t newRowLength = 0;
+
+        auto PushNewRows = [&]() {
+            extraMuseumRows.emplace_back(newRowIds, "", 0, 0);
+            extraMuseumRowColors.emplace_back(0x424242ff, 0x00000000);
+
+            newRowIds = { 0x101, 0x101, 0x101, 0x101, 0x101 };
+            newRowLength = 0;
+        };
+
+        for (auto &pair : config->tickflows) {
+            newRowIds[newRowLength] = pair.first;
+            newRowLength += 1;
+
+            if (newRowLength == 5) {
+                PushNewRows();
+            }
+        }
+
+        if (newRowLength != 0) {
+            PushNewRows();
+        }
+
+        // concat extra rows with original rows
+        std::vector<MuseumRow> originalMuseumRows = std::move(museumRows);
+        museumRows.reserve(originalMuseumRows.size() + extraMuseumRows.size());
+        museumRows.insert(museumRows.end(), extraMuseumRows.begin(), extraMuseumRows.end());
+        museumRows.insert(museumRows.end(), originalMuseumRows.begin(), originalMuseumRows.end());
+
+        // concat extra row colors with original row colors
+        std::vector<MuseumRowColor> originalMuseumRowColors = std::move(museumRowColors);
+        museumRowColors.reserve(originalMuseumRowColors.size() + extraMuseumRowColors.size());
+        museumRowColors.insert(
+            museumRowColors.end(),
+            extraMuseumRowColors.begin(),
+            extraMuseumRowColors.end()
+        );
+        museumRowColors.insert(
+            museumRowColors.end(),
+            originalMuseumRowColors.begin(),
+            originalMuseumRowColors.end()
+        );
+    }
+
     void PatchMuseumExtraRows() {
+        AddExtraRowsToFront();
+
+        // patch in new museum rows
         for (auto address : Region::MuseumRowsInfoAddresses()) {
-            Process::Patch(address, (u32) MuseumRows);
+            Process::Patch(address, (u32) museumRows.data());
         }
 
         u32 compare_r1_instruction = // cmp r1, MUSEUM_ROW_COUNT
-            make_cmp_immediate_instruction(1, MuseumRowCount);
+            make_cmp_immediate_instruction(1, museumRows.size());
 
         for (auto address : Region::MuseumRowsR1Cmps()) {
             Process::Patch(address, compare_r1_instruction);
         }
 
-        u32 compare_r8_instruction = // cmp r1, MUSEUM_ROW_COUNT
-            make_cmp_immediate_instruction(1, MuseumRowCount);
+        u32 compare_r8_instruction = // cmp r8, MUSEUM_ROW_COUNT
+            make_cmp_immediate_instruction(8, museumRows.size());
 
         for (auto address : Region::MuseumRowsR8Cmps()) {
             Process::Patch(address, compare_r8_instruction);
         }
 
-        // FIXME: for this to work we need to stop c++ from overwriting the
-        // colors when it runs the constructor (@ 0x38DE58) for the array
+        Process::Patch(Region::MuseumRowsColorsInitFunc(), /* bx lr */ 0xe12fff1e);
```

Comment:
consider either:
1. making a constant `#define BX_LR 0xe12fff1e`
2. using a hook to patch this that leads to a thunked function such as `void InitMuseumRowColors() {}`

there's not necessarily a problem with what's currently done but i think alternatives should be discussed

Reply:
I wouldn't of thought about that. letting the compiler generate the assembly is preferred

Reply:
since I didn't need the enable/disable functionality I opted to use rt's `rtGenerateJumpCode` directly. Result is the same, the  `InitMuseumRowColors` function gets redirected to a stubbed function.

Reply:
i'm honestly not sure how to feel about this. i guess it wouldn't make a lot of sense to disable this hook at runtime since it only runs once, but, idk, feels weird to have this one hook be different than the others
can you at least move StubFunction to Hooks.cpp?

Reply:
sure, should I make it generic for any `typename T`? or just keep it as a stub for void functions

Reply:
might be nice to make it generic, yeah, although i haven't really worked with generics on rust so can't really comment there

Reply:
ok, I moved it into the Hooks namespace.

I choose to have each type used with the functions require them to be instantiated first since it's simpler than having the the templated functions in headers.

Reply:
looks good